### PR TITLE
NN-3250:

### PR DESCRIPTION
**Pull request by JamieDraperUK**
- remove high complexity of needs offenders from the list of unallocated
  - rename complexity event listener to be more representative of the underlying queue
  - pull out the event processing into it's own service to remove the circular dependency

### CODE REVIEW

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerService.java [hunk 1]
```diff
@@ -131,19 +133,35 @@ private static String getKeyWorkerFullName(final KeyworkerDto keyworkerDto) {
 
 
     public List<OffenderLocationDto> getUnallocatedOffenders(final String prisonId, final String sortFields, final SortOrder sortOrder) {
-
         prisonSupportedService.verifyPrisonMigrated(prisonId);
-        final var allOffenders = nomisService.getOffendersAtLocation(prisonId, sortFields, sortOrder, false);
-        return processor.filterByUnallocated(allOffenders);
+
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerService.java:137]: New code.

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerService.java [hunk 2]
```diff
@@ -74,8 +74,10 @@
     private final KeyworkerAllocationProcessor processor;
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerService.java:74]: Mostly formatting, the new code has been marked below.

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerService.java [hunk 3]
```diff
@@ -131,19 +133,35 @@ private static String getKeyWorkerFullName(final KeyworkerDto keyworkerDto) {
...
+
+        final var offenders = removeOffendersWithHighComplexityOfNeed(
+            prisonId,
+            nomisService.getOffendersAtLocation(prisonId, sortFields, sortOrder, false)
+        );
+
+        return  processor.filterByUnallocated(offenders);
+    }
+
+    private List<OffenderLocationDto> removeOffendersWithHighComplexityOfNeed(final String prisonId, final List<OffenderLocationDto> unAllocated) {
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerService.java:146]: Bit of duplication here will look at pulling it out at a later stage

**Review by sp-watson — APPROVED** [review]

File: src/main/java/uk/gov/justice/digital/hmpps/keyworker/events/DomainEventListener.kt [hunk 4]
```diff
@@ -6,13 +6,11 @@ import org.springframework.beans.factory.annotation.Qualifier
...
 import org.springframework.jms.annotation.JmsListener
 import org.springframework.stereotype.Service
-import uk.gov.justice.digital.hmpps.keyworker.services.ComplexityOfNeed
-import java.time.LocalDateTime
 
 @Service
 @ConditionalOnExpression("{'aws', 'localstack'}.contains('\${complexity-of-need-sqs.provider}')")
-class ComplexityOfNeedListener(
-  private val complexityOfNeedService: ComplexityOfNeed,
+class DomainEventListener(
```
[the remark this answers is not in the document]
Comment (JamieDraperUK) [reply src/main/java/uk/gov/justice/digital/hmpps/keyworker/events/DomainEventListener.kt:12]: The main reason is that technically it's a Domain event queue that backs it. The queue that backs the offender events one is in fact offender events.

**Pull request merged**